File: minigit-3/py/commands/diff.py
"""diff command - Show changes."""

import sys
import os
from pathlib import Path


def run(args: list[str]) -> int:
    """Show changes."""
    from utils import find_repo_root, is_binary_file
    from index import read_index
    from refs import resolve_ref, get_head_sha
    from objects import read_commit, read_tree, read_blob
    from diff_algo import diff_files, diff_binary

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    cached = False
    staged = False
    stat_only = False
    paths = []
    commits = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '--cached':
            cached = True
        elif arg == '--staged':
            staged = True
        elif arg == '--stat':
            stat_only = True
        elif arg == '--':
            paths.extend(args[i + 1:])
            break
        elif not arg.startswith('-'):
            # Could be a commit or a path
            sha = resolve_ref(repo_root, arg)
            if sha:
                commits.append(sha)
            else:
                paths.append(arg)
        i += 1

    # Get index and tree files
    index_entries = {e['path']: e for e in read_index(repo_root)}

    head_sha = get_head_sha(repo_root)
    head_files = {}
    if head_sha:
        def get_tree_files(tree_sha: str, prefix: str = '') -> dict:
            files = {}
            try:
                tree_entries = read_tree(repo_root, tree_sha)
                for entry in tree_entries:
                    path = f"{prefix}{entry.name}" if prefix else entry.name
                    if entry.mode == '40000':
                        files.update(get_tree_files(entry.sha, path + '/'))
                    else:
                        files[path] = entry.sha
            except:
                pass
            return files

        try:
            commit = read_commit(repo_root, head_sha)
            head_files = get_tree_files(commit['tree'])
        except:
            pass

    # Determine what to diff
    output = []

    if len(commits) == 2:
        # Diff between two commits
        old_sha, new_sha = commits
        old_commit = read_commit(repo_root, old_sha)
        new_commit = read_commit(repo_root, new_sha)

        def get_tree_files_full(tree_sha: str, prefix: str = '') -> dict:
            files = {}
            try:
                tree_entries = read_tree(repo_root, tree_sha)
                for entry in tree_entries:
                    path = f"{prefix}{entry.name}" if prefix else entry.name
                    if entry.mode == '40000':
                        files.update(get_tree_files_full(entry.sha, path + '/'))
                    else:
                        files[path] = entry.sha
            except:
                pass
            return files

        old_files = get_tree_files_full(old_commit['tree'])
        new_files = get_tree_files_full(new_commit['tree'])

        all_paths_set = set(old_files.keys()) | set(new_files.keys())
        if paths:
            all_paths_set = {p for p in all_paths_set if p in paths}

        for path in sorted(all_paths_set):
            old_blob_sha = old_files.get(path)
            new_blob_sha = new_files.get(path)

            if old_blob_sha == new_blob_sha:
                continue

            old_content = None
            new_content = None

            if old_blob_sha:
                try:
                    data = read_blob(repo_root, old_blob_sha)
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    old_content = data.decode()
                except:
                    pass

            if new_blob_sha:
                try:
                    data = read_blob(repo_root, new_blob_sha)
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    new_content = data.decode()
                except:
                    pass

            diff_output = diff_files(old_content, new_content, path)
            if diff_output:
                output.append(diff_output)

    elif len(commits) == 1:
        # Diff working tree against commit
        commit_sha = commits[0]
        commit = read_commit(repo_root, commit_sha)

        def get_tree_files_full(tree_sha: str, prefix: str = '') -> dict:
            files = {}
            try:
                tree_entries = read_tree(repo_root, tree_sha)
                for entry in tree_entries:
                    path = f"{prefix}{entry.name}" if prefix else entry.name
                    if entry.mode == '40000':
                        files.update(get_tree_files_full(entry.sha, path + '/'))
                    else:
                        files[path] = entry.sha
            except:
                pass
            return files

        commit_files = get_tree_files_full(commit['tree'])

        # Get working tree files
        working_files = set()
        for root, dirs, files in os.walk(repo_root):
            if '.minigit' in dirs:
                dirs.remove('.minigit')
            for file in files:
                abs_path = Path(root) / file
                rel_path = str(abs_path.relative_to(repo_root))
                working_files.add(rel_path)

        all_paths_set = set(commit_files.keys()) | working_files
        if paths:
            all_paths_set = {p for p in all_paths_set if p in paths}

        for path in sorted(all_paths_set):
            old_sha = commit_files.get(path)
            full_path = repo_root / path

            old_content = None
            new_content = None

            if old_sha:
                try:
                    data = read_blob(repo_root, old_sha)
                    if is_binary_file(data):
                        if full_path.exists():
                            output.append(f"Binary file {path} differs")
                        continue
                    old_content = data.decode()
                except:
                    pass

            if full_path.exists():
                try:
                    data = full_path.read_bytes()
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    new_content = data.decode()
                except:
                    pass

            if old_content == new_content:
                continue

            diff_output = diff_files(old_content, new_content, path)
            if diff_output:
                output.append(diff_output)

    elif cached or staged:
        # Diff index against HEAD
        all_paths_set = set(index_entries.keys()) | set(head_files.keys())
        if paths:
            all_paths_set = {p for p in all_paths_set if p in paths}

        for path in sorted(all_paths_set):
            index_sha = index_entries.get(path, {}).get('sha')
            head_sha_val = head_files.get(path)

            if index_sha == head_sha_val:
                continue

            old_content = None
            new_content = None

            if head_sha_val:
                try:
                    data = read_blob(repo_root, head_sha_val)
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    old_content = data.decode()
                except:
                    pass

            if index_sha:
                try:
                    data = read_blob(repo_root, index_sha)
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    new_content = data.decode()
                except:
                    pass

            diff_output = diff_files(old_content, new_content, path)
            if diff_output:
                output.append(diff_output)

    else:
        # Diff working tree against index
        all_paths_set = set(index_entries.keys())

        # Get working tree files
        for root, dirs, files in os.walk(repo_root):
            if '.minigit' in dirs:
                dirs.remove('.minigit')
            for file in files:
                abs_path = Path(root) / file
                rel_path = str(abs_path.relative_to(repo_root))
                all_paths_set.add(rel_path)

        if paths:
            all_paths_set = {p for p in all_paths_set if p in paths}

        for path in sorted(all_paths_set):
            index_entry = index_entries.get(path)
            full_path = repo_root / path

            if index_entry is None:
                continue  # Untracked file

            index_sha = index_entry['sha']

            old_content = None
            new_content = None

            try:
                data = read_blob(repo_root, index_sha)
                if is_binary_file(data):
                    if full_path.exists():
                        new_data = full_path.read_bytes()
                        if data != new_data:
                            output.append(f"Binary file {path} differs")
                    continue
                old_content = data.decode()
            except:
                pass

            if full_path.exists():
                try:
                    data = full_path.read_bytes()
                    if is_binary_file(data):
                        output.append(f"Binary file {path} differs")
                        continue
                    new_content = data.decode()
                except:
                    pass
            else:
                new_content = None

            if old_content == new_content:
                continue

            diff_output = diff_files(old_content, new_content, path)
            if diff_output:
                output.append(diff_output)

    # Print output
    if stat_only:
        # Show stat format with insertions/deletions
        for diff_text in output:
            lines = diff_text.split('\n')
            path = None
            insertions = 0
            deletions = 0
            for line in lines:
                if line.startswith('diff --git'):
                    parts = line.split()
                    if len(parts) >= 4:
                        path = parts[2].lstrip('a/')
                elif line.startswith('+') and not line.startswith('+++'):
                    insertions += 1
                elif line.startswith('-') and not line.startswith('---'):
                    deletions += 1
            if path:
                changes = '+' * min(insertions, 30) + '-' * min(deletions, 30)
                total = insertions + deletions
                print(f" {path} | {total} {changes}")
    else:
        for diff_text in output:
            print(diff_text)

    return 0

File: minigit-3/ts/src/commands/diff.ts
// diff command - Show changes

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot } from '../utils';
import { readIndex } from '../index-file';
import { readObject, parseCommit, parseTree } from '../objects';
import { getHeadCommit, resolveRef } from '../refs';
import { diffFiles, formatDiff } from '../diff-algo';
import { FileDiff } from '../types';

export function diff(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let staged = false;
  let showStat = false;
  const commits: string[] = [];
  const paths: string[] = [];
  let inPathMode = false;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--') {
      inPathMode = true;
    } else if (inPathMode) {
      paths.push(arg);
    } else if (arg === '--staged' || arg === '--cached') {
      staged = true;
    } else if (arg === '--stat') {
      showStat = true;
    } else if (!arg.startsWith('-')) {
      // Could be a commit or a path
      const sha = resolveRef(arg, repoRoot);
      if (sha) {
        commits.push(sha);
      } else if (fs.existsSync(path.join(repoRoot, arg))) {
        paths.push(arg);
      } else {
        commits.push(arg); // Let it fail later
      }
    }
  }

  let diffs: FileDiff[];

  if (commits.length === 2) {
    // diff <commit1> <commit2>
    diffs = diffCommits(commits[0], commits[1], repoRoot, paths);
  } else if (commits.length === 1) {
    // diff <commit> - compare working tree to commit
    diffs = diffCommitToWorking(commits[0], repoRoot, paths);
  } else if (staged) {
    // diff --staged - compare index to HEAD
    diffs = diffIndexToHead(repoRoot, paths);
  } else {
    // diff - compare working tree to index
    diffs = diffWorkingToIndex(repoRoot, paths);
  }

  if (showStat) {
    printDiffStat(diffs);
  } else {
    printDiffs(diffs);
  }
}

function diffWorkingToIndex(repoRoot: string, filterPaths: string[]): FileDiff[] {
  const indexEntries = readIndex(repoRoot);
  const diffs: FileDiff[] = [];

  for (const entry of indexEntries) {
    if (filterPaths.length > 0 && !filterPaths.some(p => entry.name.startsWith(p) || entry.name === p)) {
      continue;
    }

    const workPath = path.join(repoRoot, entry.name);

    if (!fs.existsSync(workPath)) {
      // File deleted in working tree
      const obj = readObject(entry.sha, repoRoot);
      const diff = diffFiles(obj.content.toString(), '', {
        oldPath: entry.name,
        newPath: entry.name,
        oldMode: entry.mode.toString(8).padStart(6, '0')
      });
      diff.isDeleted = true;
      diffs.push(diff);
    } else {
      const stats = fs.statSync(workPath);
      if (stats.isFile()) {
        const workContent = fs.readFileSync(workPath, 'utf-8');
        const obj = readObject(entry.sha, repoRoot);
        const indexContent = obj.content.toString();

        if (workContent !== indexContent) {
          const diff = diffFiles(indexContent, workContent, {
            oldPath: entry.name,
            newPath: entry.name,
            oldMode: entry.mode.toString(8).padStart(6, '0'),
            newMode: entry.mode.toString(8).padStart(6, '0')
          });
          diffs.push(diff);
        }
      }
    }
  }

  return diffs;
}

function diffIndexToHead(repoRoot: string, filterPaths: string[]): FileDiff[] {
  const indexEntries = readIndex(repoRoot);
  const headCommit = getHeadCommit(repoRoot);
  const diffs: FileDiff[] = [];

  // Get HEAD tree files
  const headFiles = new Map<string, { sha: string; mode: string }>();
  if (headCommit) {
    const commitObj = readObject(headCommit, repoRoot);
    const commit = parseCommit(commitObj.content);
    collectTreeFilesWithMode(commit.tree, '', repoRoot, headFiles);
  }

  const indexMap = new Map(indexEntries.map(e => [e.name, e]));

  // Check files in index
  for (const entry of indexEntries) {
    if (filterPaths.length > 0 && !filterPaths.some(p => entry.name.startsWith(p) || entry.name === p)) {
      continue;
    }

    const headEntry = headFiles.get(entry.name);

    if (!headEntry) {
      // New file
      const obj = readObject(entry.sha, repoRoot);
      const diff = diffFiles('', obj.content.toString(), {
        oldPath: entry.name,
        newPath: entry.name,
        newMode: entry.mode.toString(8).padStart(6, '0')
      });
      diff.isNew = true;
      diffs.push(diff);
    } else if (headEntry.sha !== entry.sha) {
      // Modified file
      const headObj = readObject(headEntry.sha, repoRoot);
      const indexObj = readObject(entry.sha, repoRoot);
      const diff = diffFiles(headObj.content.toString(), indexObj.content.toString(), {
        oldPath: entry.name,
        newPath: entry.name,
        oldMode: headEntry.mode,
        newMode: entry.mode.toString(8).padStart(6, '0')
      });
      diffs.push(diff);
    }
  }

  // Check files deleted from index
  for (const [name, headEntry] of headFiles) {
    if (filterPaths.length > 0 && !filterPaths.some(p => name.startsWith(p) || name === p)) {
      continue;
    }

    if (!indexMap.has(name)) {
      const headObj = readObject(headEntry.sha, repoRoot);
      const diff = diffFiles(headObj.content.toString(), '', {
        oldPath: name,
        newPath: name,
        oldMode: headEntry.mode
      });
      diff.isDeleted = true;
      diffs.push(diff);
    }
  }

  return diffs;
}

function diffCommits(sha1: string, sha2: string, repoRoot: string, filterPaths: string[]): FileDiff[] {
  const commit1 = parseCommit(readObject(sha1, repoRoot).content);
  const commit2 = parseCommit(readObject(sha2, repoRoot).content);

  const files1 = new Map<string, { sha: string; mode: string }>();
  const files2 = new Map<string, { sha: string; mode: string }>();

  collectTreeFilesWithMode(commit1.tree, '', repoRoot, files1);
  collectTreeFilesWithMode(commit2.tree, '', repoRoot, files2);

  const diffs: FileDiff[] = [];
  const allPaths = new Set([...files1.keys(), ...files2.keys()]);

  for (const name of allPaths) {
    if (filterPaths.length > 0 && !filterPaths.some(p => name.startsWith(p) || name === p)) {
      continue;
    }

    const entry1 = files1.get(name);
    const entry2 = files2.get(name);

    if (!entry1) {
      // New file in commit2
      const obj = readObject(entry2!.sha, repoRoot);
      const diff = diffFiles('', obj.content.toString(), {
        oldPath: name,
        newPath: name,
        newMode: entry2!.mode
      });
      diff.isNew = true;
      diffs.push(diff);
    } else if (!entry2) {
      // Deleted in commit2
      const obj = readObject(entry1.sha, repoRoot);
      const diff = diffFiles(obj.content.toString(), '', {
        oldPath: name,
        newPath: name,
        oldMode: entry1.mode
      });
      diff.isDeleted = true;
      diffs.push(diff);
    } else if (entry1.sha !== entry2.sha) {
      // Modified
      const obj1 = readObject(entry1.sha, repoRoot);
      const obj2 = readObject(entry2.sha, repoRoot);
      const diff = diffFiles(obj1.content.toString(), obj2.content.toString(), {
        oldPath: name,
        newPath: name,
        oldMode: entry1.mode,
        newMode: entry2.mode
      });
      diffs.push(diff);
    }
  }

  return diffs;
}

function diffCommitToWorking(sha: string, repoRoot: string, filterPaths: string[]): FileDiff[] {
  const commit = parseCommit(readObject(sha, repoRoot).content);
  const commitFiles = new Map<string, { sha: string; mode: string }>();
  collectTreeFilesWithMode(commit.tree, '', repoRoot, commitFiles);

  const workingFiles = new Set<string>();
  collectWorkingFiles(repoRoot, '', repoRoot, workingFiles);

  const diffs: FileDiff[] = [];
  const allPaths = new Set([...commitFiles.keys(), ...workingFiles]);

  for (const name of allPaths) {
    if (filterPaths.length > 0 && !filterPaths.some(p => name.startsWith(p) || name === p)) {
      continue;
    }

    const commitEntry = commitFiles.get(name);
    const workPath = path.join(repoRoot, name);
    const existsInWorking = fs.existsSync(workPath);

    if (!commitEntry && existsInWorking) {
      // New file
      const content = fs.readFileSync(workPath, 'utf-8');
      const diff = diffFiles('', content, { oldPath: name, newPath: name });
      diff.isNew = true;
      diffs.push(diff);
    } else if (commitEntry && !existsInWorking) {
      // Deleted
      const obj = readObject(commitEntry.sha, repoRoot);
      const diff = diffFiles(obj.content.toString(), '', {
        oldPath: name,
        newPath: name,
        oldMode: commitEntry.mode
      });
      diff.isDeleted = true;
      diffs.push(diff);
    } else if (commitEntry && existsInWorking) {
      const workContent = fs.readFileSync(workPath, 'utf-8');
      const commitObj = readObject(commitEntry.sha, repoRoot);
      const commitContent = commitObj.content.toString();

      if (workContent !== commitContent) {
        const diff = diffFiles(commitContent, workContent, {
          oldPath: name,
          newPath: name,
          oldMode: commitEntry.mode
        });
        diffs.push(diff);
      }
    }
  }

  return diffs;
}

function collectTreeFilesWithMode(treeSha: string, prefix: string, repoRoot: string, files: Map<string, { sha: string; mode: string }>): void {
  const treeObj = readObject(treeSha, repoRoot);
  const entries = parseTree(treeObj.content);

  for (const entry of entries) {
    const name = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.type === 'tree') {
      collectTreeFilesWithMode(entry.sha, name, repoRoot, files);
    } else {
      files.set(name, { sha: entry.sha, mode: entry.mode });
    }
  }
}

function collectWorkingFiles(dir: string, prefix: string, repoRoot: string, files: Set<string>): void {
  if (!fs.existsSync(dir)) return;

  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    if (entry.name === '.minigit') continue;

    const name = prefix ? `${prefix}/${entry.name}` : entry.name;

    if (entry.isDirectory()) {
      collectWorkingFiles(path.join(dir, entry.name), name, repoRoot, files);
    } else if (entry.isFile()) {
      files.add(name);
    }
  }
}

function printDiffs(diffs: FileDiff[]): void {
  for (const diff of diffs) {
    console.log(formatDiff(diff));
  }
}

function printDiffStat(diffs: FileDiff[]): void {
  let totalInsertions = 0;
  let totalDeletions = 0;

  for (const diff of diffs) {
    let insertions = 0;
    let deletions = 0;

    for (const hunk of diff.hunks) {
      for (const line of hunk.lines) {
        if (line.startsWith('+') && !line.startsWith('+++')) {
          insertions++;
        } else if (line.startsWith('-') && !line.startsWith('---')) {
          deletions++;
        }
      }
    }

    const changes = insertions + deletions;
    const bar = '+'.repeat(Math.min(insertions, 30)) + '-'.repeat(Math.min(deletions, 30));
    console.log(` ${diff.newPath.padEnd(40)} | ${String(changes).padStart(4)} ${bar}`);

    totalInsertions += insertions;
    totalDeletions += deletions;
  }

  if (diffs.length > 0) {
    console.log(` ${diffs.length} file${diffs.length > 1 ? 's' : ''} changed, ${totalInsertions} insertion${totalInsertions !== 1 ? 's' : ''}(+), ${totalDeletions} deletion${totalDeletions !== 1 ? 's' : ''}(-)`);
  }
}
